clients/src/main/java/org/apache/kafka/common/utils/MappedByteBuffers.java

```diff
@@ -0,0 +1,139 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+
+import static java.lang.invoke.MethodHandles.constant;
+import static java.lang.invoke.MethodHandles.dropArguments;
+import static java.lang.invoke.MethodHandles.filterReturnValue;
+import static java.lang.invoke.MethodHandles.guardWithTest;
+import static java.lang.invoke.MethodHandles.lookup;
+import static java.lang.invoke.MethodType.methodType;
+
+/**
+ * Utility methods for MappedByteBuffer implementations.
+ *
+ * The unmap implementation was inspired by the one in Lucene's MMapDirectory.
+ */
+public final class MappedByteBuffers {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MappedByteBuffers.class);
+
+    // null if unmap is not supported
+    private static final MethodHandle UNMAP;
+
+    // null if unmap is supported
+    private static final String UNMAP_NOT_SUPPORTED_REASON;
+
+    static {
+        Object unmap = lookupUnmapMethodHandle();
+        if (unmap instanceof MethodHandle) {
+            UNMAP = (MethodHandle) unmap;
+            UNMAP_NOT_SUPPORTED_REASON = null;
+        } else {
+            UNMAP = null;
+            UNMAP_NOT_SUPPORTED_REASON = (String) unmap;
+        }
+    }
+
+    private MappedByteBuffers() {}
+
+    public static void unmap(String resourceDescription, MappedByteBuffer buffer) throws IOException {
+        if (!buffer.isDirect())
+            throw new IllegalArgumentException("Unmapping only works with direct buffers");
+        if (UNMAP == null)
+            throw new UnsupportedOperationException(UNMAP_NOT_SUPPORTED_REASON);
+
+        try {
+            UNMAP.invokeExact((ByteBuffer) buffer);
+        } catch (Throwable throwable) {
+            throw new IOException("Unable to unmap the mapped buffer: " + resourceDescription, throwable);
+        }
+    }
+
+    /**
+     * Return an unmap MethodHandle or a String with the reason why unmap is not supported.
+     */
+    private static Object lookupUnmapMethodHandle() {
+        final MethodHandles.Lookup lookup = lookup();
+        try {
+            try {
+                return unmapJava9(lookup);
+            } catch (SecurityException e) {
+                return e.getMessage();
+            } catch (ReflectiveOperationException | RuntimeException e) {
+                LOGGER.debug("Falling back to unmap implementation for Java 7/Java 8 due to " + e.getMessage());
+                return unmapJava7Or8(lookup);
+            }
+        } catch (SecurityException se) {
+            return se.getMessage();
```

Comment:
Rather an odd thing to do, but I suppose it is ok since it is private and there is no easier way to set the final message.

Reply:
Yeah, I got this pattern from the Lucene code and it's super weird. The straightforward try/catch doesn't work if we want to keep the static fields final. However, by using some locals we can achieve the goal without this super weird pattern. I'll push an update.